sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +308,103 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
     }
 
     @VisibleForTesting
-    protected IndexItemResponse indexItem(ShardUpsertRequest request,
-                                          ShardUpsertRequest.Item item,
-                                          IndexShard indexShard,
-                                          boolean tryInsertFirst,
-                                          UpdateSourceGen updateSourceGen,
-                                          InsertSourceGen insertSourceGen,
-                                          @Nullable ReturnValueGen returnGen,
-                                          boolean isRetry) throws Exception {
-        final long seqNo;
-        final long primaryTerm;
-        final long version;
-        Doc updatedDoc = null;
-        if (tryInsertFirst) {
-            version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE
-                ? Versions.MATCH_ANY
-                : Versions.MATCH_DELETED;
-            seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
-            primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
-            try {
-                item.source(insertSourceGen.generateSourceAndCheckConstraints(item.insertValues()));
-            } catch (IOException e) {
-                throw ExceptionsHelper.convertToElastic(e);
+    protected IndexItemResponse insert(ShardUpsertRequest request,
+                                       ShardUpsertRequest.Item item,
+                                       IndexShard indexShard,
+                                       boolean isRetry,
+                                       @Nullable ReturnValueGen returnGen,
+                                       @Nullable InsertSourceGen insertSourceGen) throws Exception {
+        assert insertSourceGen != null : "InsertSourceGen must not be null";
+        final BytesReference rawSource;
+        Map<String, Object> source = null;
+        try {
+            if (insertSourceGen instanceof FromRawInsertSource) {
```

Comment:
Optimize here for the case where to insert value is already string-based.

Reply:
I think it would be worth adding the information that this is primarily a performance optimization as inline comment.